pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/batching/AbstractBatchedMetadataStore.java

```diff
@@ -0,0 +1,162 @@
+        } else {
+            scheduledTask = null;
+            readOps = null;
+            writeOps = null;
+        }
+    }
+
+    @Override
+    public void close() throws Exception {
+        if (enabled) {
+            // Fail all the pending items
+            Exception ex = new IllegalStateException("Metadata store is getting closed");
+            readOps.drain(op -> op.getFuture().completeExceptionally(ex));
+            writeOps.drain(op -> op.getFuture().completeExceptionally(ex));
+
+            scheduledTask.cancel(true);
+        }
+        super.close();
+    }
+
+    private void flush() {
+        while (!readOps.isEmpty()) {
+            List<MetadataOp> ops = new ArrayList<>();
+            readOps.drain(ops::add, maxOperations);
+            batchOperation(ops);
+        }
+
+        while (!writeOps.isEmpty()) {
+            int batchSize = 0;
+
+            List<MetadataOp> ops = new ArrayList<>();
+            for (int i = 0; i < maxOperations; i++) {
+                MetadataOp op = writeOps.peek();
+                if (op == null) {
+                    break;
+                }
+
+                if (i > 0 && (batchSize + op.size()) > maxSize) {
+                    // We have already reached the max size, so flush the current batch
+                    break;
+                }
+
+                batchSize += op.size();
+                ops.add(writeOps.poll());
+            }
+            batchOperation(ops);
+        }
+
+        flushInProgress.set(false);
+    }
+
+    @Override
+    public final CompletableFuture<Optional<GetResult>> storeGet(String path) {
+        OpGet op = new OpGet(path);
+        enqueue(readOps, op);
+        return op.getFuture();
+    }
+
+    @Override
+    protected final CompletableFuture<List<String>> getChildrenFromStore(String path) {
+        OpGetChildren op = new OpGetChildren(path);
+        enqueue(readOps, op);
+        return op.getFuture();
+    }
+
+    @Override
+    protected final CompletableFuture<Void> storeDelete(String path, Optional<Long> expectedVersion) {
+        OpDelete op = new OpDelete(path, expectedVersion);
+        enqueue(writeOps, op);
+        return op.getFuture();
+    }
+
+    @Override
+    protected final CompletableFuture<Stat> storePut(String path, byte[] data, Optional<Long> optExpectedVersion,
+                                               EnumSet<CreateOption> options) {
+        OpPut op = new OpPut(path, data, optExpectedVersion, options);
+        enqueue(writeOps, op);
+        return op.getFuture();
+    }
+
+    private void enqueue(MessagePassingQueue queue, MetadataOp op) {
+        if (enabled) {
+            if (!queue.offer(op)) {
+                // Execute individually if we're failing to enqueue
+                batchOperation(Collections.singletonList(op));
+                return;
+            }
+            if (queue.size() > maxOperations && flushInProgress.compareAndSet(false, true)) {
```

Comment:
We will not trigger a flush if reach the size limitation? Looks like the current implementation is split into multiple batches when flushing, if we have enough data should be flush not the operations, we will wait for `maxDelayMillis`.

Reply:
The change LGTM and just left a minor comment, I think it should not be a big problem here, but will simplify implementation.

Reply:
You're right, there are several optimizations that are still possible in the batching logic :) I'd say to continue in subsequent PRs.